Added privateKey additional format option

seuf opened the pull request:
**What this PR does / why we need it**:

This PR add a new `additionalOutputFormats` option in the certificates crd.
For now additional supported formats are :
- `DER` this will create a new entry `tls.der` in the kubernetes secret with the private key in DER format (binary).
- `CombindedPEM` this will create a new entry `tls.pem` in the kubernetes secret with the concatenation of the private key and the certificate.

**Which issue this PR fixes** :

 fixes  #843 (see https://github.com/jetstack/cert-manager/issues/843#issuecomment-566054175)


**Use-case : Postgres JBDC driver (lower than 42.2.8)** :

In recent versions of the Postgres JBDC driver (42.2.9 and above, which was released in Dec 2019), you can use the PKCS#12 format to load the private key and certificate. cert-manager supports PKCS#12 since 0.14, which means the integration with cert-manager works well. Postgresql 13+ [is required](https://www.postgresql.org/docs/13/libpq-connect.html#LIBPQ-CONNECT-SSLPASSWORD) to accept the `sslpassword` parameter to unlock the PKCS#12 keystore.

But if you are stuck with a version of the Postgres JDBC driver that is 42.2.9 and below, then the only way is to use the sslkey configuration field. The sslkey requires the path to a file containing the private key PKCS#8-formated and DER-encoded.

For example, I've succeffully builded and deployed this version in my cluster and used the new `key.der` private key file to connect to my postgresql database with mTLS :

```
url: jdbc:postgresql://postgresql.database.svc:5432/${postgresql.database}?ssl=true&sslmode=verify-ca&sslrootcert=/certs/ca.crt&sslcert=/certs/tls.crt&sslkey=/certs/key.der&ApplicationName=${spring.application.name}
```

With pg_hba.conf like 
```
hostssl mydatabase all 10.0.0.0/8 cert map=applications
```

and pg_ident.conf 
```
applications /^(.*)\.mynamespace\.svc$ \1
```

**Release note**:
```release-note
Added additionalOutputFormats parameter to allow `DER` (binary) and `CombinedPEM` (key + cert bundle) formats.
```

SgtCoDFish commented:
/kind feature
/ok-to-test

munnerz commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -231,6 +231,11 @@ type CertificatePrivateKey struct {
+	// If `withDerFormar` is `true`, a new entry `tls.der` will be added in the secret,
+	// containing the binary value of the private key.
+	// +optional
+	WithDerFormat bool `json:"withDerFormat,omitempty"`
```
Totally aside from whether we should or shouldn't add this feature, I think this field makes more sense as a `format` field or similar (rather than a bool option). This would also allow us to expand the supported options in future.

SgtCoDFish replied:
Fully agree with this, if we're gonna add this functionality I'd propose something like:

```suggestion
	AdditionalKeyFormats []string `json:"additionalKeyFormats,omitempty"`
```

specified like this:

```yaml
  privateKey:
    algorithm: ECDSA
    size: 256
    additionalKeyFormats: ["tls.der"]
```

where each entry in the array maps to the key which would be added to the secret. This would make it pretty easy to add other formats for the private key.

seuf replied:
updated PR to add `additionalOutputFormats`.

munnerz commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -205,6 +206,11 @@ func (s *SecretsManager) setValues(crt *cmapi.Certificate, secret *corev1.Secret
+	if crt.Spec.PrivateKey.WithDerFormat {
+		block, _ := pem.Decode(data.PrivateKey)
+		secret.Data["tls.der"] = block.Bytes
+	}
```
Does it make sense for us to supplement the existing key (tls.key), or replace it?

munnerz replied:
Also if a user were to enable this, and then later disable it, the key would continue to exist but just be out of date.

seuf replied:
I've tried to replace the actual tls.key with the DER formated key, but It failed during the truststore creation because it can't decode it again. That's why I've Just added a boolean option and keep the `tls.key` in PEM format.

SgtCoDFish replied:
I can't find a good source on whether `kubernetes.io/tls` secrets _require_ that `tls.crt` and `tls.key` be PEM formatted - [this page](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets) mentions that they should be PEM formatted when imported with `kubectl create secret tls ...` and they're certainly PEM formatted when stored after being created in that way.

That said, even if cert-manager didn't assume that `tls.key` is PEM formatted, I would vote that `tls.key` should _always_ exist and should _always_ be PEM formatted, which would imply that we'd add additional `secret.Data` keys (as in this PR) would be the preferred approach if this is something we want to implement.

SgtCoDFish replied:
> Also if a user were to enable this, and then later disable it, the key would continue to exist but just be out of date.

NB: I've confirmed this behaviour locally (I didn't doubt it was true but I wanted to check for my own benefit)

SgtCoDFish replied:
The JKS and PKCS12 keystores don't have this behaviour, and they delete the keys explicitly:

https://github.com/jetstack/cert-manager/blob/7923823cff0afe5af49282bde3a5a1cc6f5abc84/pkg/controller/certificates/internal/secretsmanager/secret.go#L203

I think this could be replicated for DER keys?

munnerz commented:
I think the strategy to implement this needs careful consideration - I highlighted a couple of cases, but you should also consider what happens if a user sets this on an existing Certificate. With this implementation, the Secret resource won't be updated until a re-issuance occurs.

A more intelligent controller could read the `tls.key` entry and actually extract the DER encoded contents out.

Have you considered building this as an entirely external controller that watches for secrets of type `kubernetes.io/tls`, reads `tls.pem` and writes `tls.der` back into the Secret? This would probably be far simpler to write (less complexity in the cert-manager codebase to deal with) as well as being entirely generic too.

munnerz commented:
I'm also conscious that #843 actually references a whole host of alternative formats, not just DER. We need to consider what the implementation plan for the others mentioned in that issue is (in terms of API surface) to ensure we don't create an 'awkward' API.

The general consensus in #843 was that it's probably easier to implement this stuff externally due to the huge number of formats (and then additional options on those formats, e.g. passwords to encrypt bundles etc)

seuf commented:
> Have you considered building this as an entirely external controller that watches for secrets of type `kubernetes.io/tls`, reads `tls.pem` and writes `tls.der` back into the Secret? This would probably be far simpler to write (less complexity in the cert-manager codebase to deal with) as well as being entirely generic too.
Do you mean a controller like https://github.com/maelvls/secret-transform ?
The idea is to watch kubernetes secrets and add new entry tls.der each time a tls.key is present..
I've considered this option but it mean a new deployment + rbac + config, etc...  This PR just add a single boolean parameter in the certificate crd and do the job :sunglasses:

munnerz commented:
> > Have you considered building this as an entirely external controller that watches for secrets of type `kubernetes.io/tls`, reads `tls.pem` and writes `tls.der` back into the Secret? This would probably be far simpler to write (less complexity in the cert-manager codebase to deal with) as well as being entirely generic too.
> 
> Do you mean a controller like https://github.com/maelvls/secret-transform ? The idea is to watch kubernetes secrets and add new entry tls.der each time a tls.key is present.. I've considered this option but it mean a new deployment + rbac + config, etc... This PR just add a single boolean parameter in the certificate crd and do the job 😎

Yep precisely 😄 the difference is that it's not quite that simple when you also account for keeping the field up to date (e.g. if the Secret ever changes), flipping the field from true->false, and also how we accommodate other secret formats. We've typically err'd on the side of allowing users to extend the functionality of the system through other controllers rather than making our own monolithic controller for all different types of transformations that may need to be applied.

SgtCoDFish commented:
## Overall

I think it would be totally valid and a good improvement to add additional keys to the secret via a `format` or `additionalKeyFormats` or whatever. I fully agree with [this comment](https://github.com/jetstack/cert-manager/pull/4598#discussion_r750067380), though, that a different name than `WithDerFormat` would be appropriate and would be easier to extend in the future if we want to add more private key types.

Crucially, the private key is information that cert-manager already has (i.e. it already knows the private key when it's writing the secret) and it doesn't seem to be a massive burden in terms of memory, CPU, storage or network requests to add this functionality. Ensuring that the cert/key is written in a format which is useful to people is a reasonable thing to add to cert-manager and is worth expanding the API for. I don't think there are enough private key formats in the wild that it would become a burden for us to add support for more. It's quite distinct from, say, ACME challenge solvers or `Issuer`s, both of which I think absolutely should be implemented using new controllers.

cert-manager writing the DER formatted secret also avoids having a "DER controller" in the cluster which has permissions to read private keys, which would increase attack surface. It also avoids having to run another thing which is valuable in and of itself, as @seuf [alluded to](https://github.com/jetstack/cert-manager/pull/4598#issuecomment-970232829) earlier.

Security of a "DER controller" aside, I feel the UX would be better if cert-manager itself wrote the DER key to the secret. Having a "DER controller" augment the secret after cert-manager writes it means that for a pod which needs the DER key, the secret won't be ready until the second controller is finished. The question "is the secret ready?" becomes harder to answer.

(An alternative approach would be that the "DER controller" creates a second secret derived from the secret which cert-manager writes, but this creates another place where one private key is stored, making audits harder and complicating things from an operational perspective)

## This PR

I think for me to be able to approve this PR, it would need:

1. A different name to `WithDerFormat`
2. At least one e2e test which validates this new behaviour
3. An answer to the issue raised in [this comment](https://github.com/jetstack/cert-manager/pull/4598#discussion_r750068463). That's a sharp edge which could catch people out; maybe a warning would be enough, but ideally the `tls.der` key would be removed.
4. Ideally, for reissuance to be triggered by `WithDerFormat`'s value being changed.

munnerz commented:
So I think the API format may want to encompass additional options for potential other formats in future too, e.g.:

```
privateKey:
    algorithm: ECDSA
    size: 256
    additionalKeyFormats:
    - type: DER
    - type: some-type-with-options
      someAdditionalOption: xyz
```

> Ideally, for reissuance to be triggered by WithDerFormat's value being changed.

This is something I'm less certain on, as the 'keystore' stuff we have does *not* do this. Right now we are purely driven by `tls.key` and `tls.crt`.

Another thing to consider that makes it harder for us to _remove_ `tls.der` from the Secret if this option is later disabled, is that some people may be running with a similar external controller to do this today, and if we remove this field we may 'fight' this controller (by having us fully manage and remove the key when not needed is effectively a breaking change to some).

SgtCoDFish commented:
> I think the API format may want to encompass additional options

An array of objects would definitely be more extensible than an array of strings, yeah; could also allow people to change the name of the key we add to the secret, and to tweak parameters (like, say, p12 PBKDF2 iterations).

> This is something I'm less certain on, as the 'keystore' stuff we have does not do this. Right now we are purely driven by tls.key and tls.crt.

Fair point, yeah! Maybe it doesn't _need_ to reissue; there's always `cmctl renew`. I'll update my above comment.

> Another thing to consider that makes it harder for us to remove tls.der...

I think it's always reasonable for us to overwrite the secret `Data` in its entirety on re-issuance. I accept and understand that it _could_ break someone but if a pod relies on a secret being augmented with an additional key by a non-cert-manager controller after the secret is written, it _must_ have logic in place to wait for the augmentation to complete or else it's always at risk to breakage.

This comes back to what I mentioned before about the question "is the secret ready?" becoming harder to answer. I think when cert-manager has written a secret after issuance, the answer to "is the secret ready?" should always be "yes" as far as cert-manager is concerned. If a user adds a custom controller which changes the answer to "maybe not yet", I think it's entirely the user's responsibility to add logic to wait for the answer to become "yes" again.

seuf commented:
> > > Have you considered building this as an entirely external controller that watches for secrets of type `kubernetes.io/tls`, reads `tls.pem` and writes `tls.der` back into the Secret? This would probably be far simpler to write (less complexity in the cert-manager codebase to deal with) as well as being entirely generic too.
> > 
> > 
> > Do you mean a controller like https://github.com/maelvls/secret-transform ? The idea is to watch kubernetes secrets and add new entry tls.der each time a tls.key is present.. I've considered this option but it mean a new deployment + rbac + config, etc... This PR just add a single boolean parameter in the certificate crd and do the job sunglasses
> 
> Yep precisely smile the difference is that it's not quite that simple when you also account for keeping the field up to date (e.g. if the Secret ever changes), flipping the field from true->false, and also how we accommodate other secret formats. We've typically err'd on the side of allowing users to extend the functionality of the system through other controllers rather than making our own monolithic controller for all different types of transformations that may need to be applied.

OK, I've forked @maelvls project to adapt it for my needs : https://github.com/seuf/secret-transform
I've builded and deployed it in my cluster, it works.

Regarding this PR I can try to adapt it to be more extensible in the future, but it will need more work.
Is it ok for you if we go with a `privateKey.additionalKeyFormats` to only convert the tls.key in the desired format (`DER` only for now) without removing it in the secret if the option is disabled (to not fight with a potential external controllers)

SgtCoDFish commented:
> Regarding this PR I can try to adapt it to be more extensible in the future, but it will need more work.

Yeah I definitely don't think we need to add other formats than DER to start off with. If you're be interested in making this change, I think it's reasonable to add `privateKey.additionalKeyFormats` something like this to start:

```yaml
                    additionalKeyFormats:
                      description: Other formats in which the key should be written to a secret
                      type: array
                      items:
                        type: object
                        properties:
                          type:
                            description: The type of additional private key format to write
                            type: string
                            enum:
                            - der
```

seuf commented:
OK, I've pushed a new version of the PR with additionalKeyFormats.
Tested in my cluster it works :sunglasses:

munnerz commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,19 @@ var (
+
+const (
+	// DER Key format is the binary format of the key
+	DERKeyFormat KeyFormat = "DER"
```
Perhaps add a note here to mention that this will be stored in a key named `tls.der` :)

seuf replied:
I've added description of create files according to each format.
I've also added the `PEM` format to create a tls.pem file containing tls.key + tls.crt (concatened)

munnerz reviewed (commented):
Can you also add some unit tests to the `secretsmanager` controller & an end-to-end test that exercises this functionality?

Adding the end-to-end test to this directory probably makes most sense: https://github.com/jetstack/cert-manager/blob/master/test/e2e/suite/issuers/selfsigned/certificate.go (we only need to exercise this codepath for one issuer, and the selfsigned issuer is nice and lightweight 😄)

SgtCoDFish commented on internal/apis/certmanager/types_certificate.go:
```diff
@@ -203,6 +203,25 @@ type CertificatePrivateKey struct {
+	// stored in `tls.der` secret key
+	DERKeyFormat KeyFormat = "DER"
+
+	// PEM Key format is a file containing both tls.crt and tls.key
+	// stored in `tls.pem` secret key
+	PEMKeyFormat KeyFormat = "PEM"
```
suggestion: I think it'd be best if we keep this PR _only_ for adding the DER-encoded private key. The combined PEM is a good idea which I'd support, but I think it's best to minimise the change in any given PR to make it easier to review.

I think adding the combined PEM in a separate PR would be quick and painless when / if this PR merges.

seuf replied:
As discussed on slack the feature is now flagged.
So it should not impact users to add those two additional Key Formats.
Also the PEM formating is a very small part of this PR :sweat_smile:

munnerz replied:
This has actually raised a few questions for me:

1) The field is called `additionaKeyFormats`, however, the PEM option here encodes a certificate as well, so 'key' here isn't correct/clear.

2) It _is_ possible to encode a certificate in DER format too, not just a private key (but we can't concatenate them into a single file as _that_ is part of the PEM spec, not DER). If users want DER private keys, is it reasonable to expect they want DER certificate data?

3) Should we therefore move this to `spec.additionalOutputFormats`?

4) If so, and we _do_ include a `CombinedPEM` option, would JKS keystores or PKCS#12 keystores also be configured in a similar way?

I think we need to reconsider how we're expressing these options to make sure we create an API that is cohesive/consistent between different formats (and clear to users when they use it)

cc @SgtCoDFish @maelvls @JoshVanL @irbekrm

SgtCoDFish replied:
Your questions are super insightful @munnerz !

There's actually a bunch of overlap here with the `spec.keystores` field on the Certificate resource... so much so, that I can't really think of a reason that we couldn't add DER keys to `keystores` (i.e. add a new field [here](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1.CertificateKeystores)). Combined chain + key PEM files would also fit under there.

I think both "DER private key" and "combined PEM" would fit under the name "keystores" too. A DER-encoded certificate probably wouldn't fit under "keystores", but I think we could kick that can down the road since nobody's asking for that.

For example:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  ...
spec:
  ...
  keystores:
    derPrivateKey:
      create: true
      # can't think of any configurable options we might want here but we can add later if wanted
    combinedPEM: # combinedPEM just an example, I still don't think we should do this in this PR
      create: true
      ordering: enum(chainFirst | keyFirst) # an example of a configurable option we might want
  ...
```

munnerz replied:
hmm, I disagree - combinedPEM is not a type of keystore, and at least when that field was added it was intended specifically because PKCS#12 and JKS are specifically seen as 'keystores' in common language in the area.

I was more inclined to move `keystores` to be a type of option under `spec.additionalOutputFormats`, e.g.:

```
spec:
  additionalOutputFormats:
  - type: JKS
    passwordSecretRef: a-secret
  - type: DER
  - type: PKCS12
    passwordSecretRef: pkcs12-password
  - type: CombinedPEM
```

munnerz replied:
(but I do agree with your other points regarding this being closely related! 😄)

SgtCoDFish replied:
> I was more inclined to move keystores to be a type of option under spec.additionalOutputFormats...

Sure, in an ideal world I fully agree; in practice though this also requires either a migration / deprecation process or for us to support JKS/PKCS12 keystores being requested in two separate places, while also requiring us to write more code to support and test an entirely new field with a new syntax.

I don't think any reasonable person is going to struggle with accepting "keystores" being defined as "any format where a private key is stored"

Under that definition, we wouldn't even need to update the doc comments for `CertificateKeystores` to add both DER and "combined-PEM":

> CertificateKeystores configures additional keystore output formats to be created in the Certificate’s output Secret

SgtCoDFish replied:
> I am much more in favor of the additionalOutputFormats option which 1. clearly describes what's going on ("I want additional output formats with my signed certificate chain/private key data/certificate authority data"), and 2. gives scope for adding additional fields to each element, or weird and wacky output formats users might want to use in the future.

In an ideal world I agree and I'd agree for `Certificate` v2 if it's ever created. I still think keystore would be far less confusing for people than having a different field, so that we don't end up with two different fields on the `Certificate` resource which both mean "add some data to my secret under a different key".

I think anything containing a private key could reasonably be thought of as a keystore. JKS and PKCS12 files both contain certificates and private keys and we don't have a problem calling them keystores. If the request was just for a PEM _cert_ with no key I could see why it would be confusing.

That said this _really_ doesn't matter and endlessly bikeshedding API design is a surefire way to never accomplish anything. If adding a new field to `v1.Certificate` will get this (really basic) feature merged then I'll support adding `additionalOutputFormats` [1]. 

[1] (but I wouldn't support then adding JKS and PKCS12 to `additionalOutputFormats` down the road for `v1.Certificate`, that creates two ways of achieving identical outcomes, doubling testing and code attack surface.)

SgtCoDFish replied:
@seuf 

In addition, I think it bears repeating here that making this PR _only_ do DER keys would make it much easier to merge, which was the purpose of my original comment on this thread:

> suggestion: I think it'd be best if we keep this PR only for adding the DER-encoded private key. The combined PEM is a good idea which I'd support, but I think it's best to minimise the change in any given PR to make it easier to review.

munnerz replied:
> That said this really doesn't matter and endlessly bikeshedding API design is a surefire way to never accomplish anything.

Without going too off-topic on this PR, I somewhat disagree with this... the API is the way the user interacts with our software, and it is a difficult part to _change_ once we've put it out to the world. Feature gating does help, but historically I don't think we've got a good track record of ownership on a feature-gated feature and we've not been so good at reflecting & iterating on the design later down the road.

I think it's good for us to have this conversation, and especially _with_ the PR author, so that we all have a shared/common understanding of what we're trying to achieve and the longer-term plan for the feature's graduation :)

But I don't want to drag this thread aside from the point at hand, so let's chat on Slack if we want to follow up/bikeshed on our bikeshedding ;)

SgtCoDFish replied:
I'll rephrase it:

I think adding this to `keystores` would be better, you both disagree with me. It really, _really_ doesn't matter in the grand scheme of things [1] so let's add `additionalOutputFormats` rather than have me spend time advocating for `keystores`. I'm saying I'm not going to be a blocker; I want to get something done.

[1] The finer points of API design aren't lost on me and I'm well aware of the difficulties of changing a public API but we're just discussing whether to refer to `.spec.keystores.der` or `spec.additionalOutputFormats[].type = der`, it's two different ways of naming the new thing we're going to add.

SgtCoDFish commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -203,8 +204,22 @@ func (s *SecretsManager) setValues(crt *cmapi.Certificate, secret *corev1.Secret
+				switch f.Type {
+				case cmapi.DERKeyFormat:
+					block, _ := pem.Decode(data.PrivateKey)
+					secret.Data["tls.der"] = block.Bytes
```
suggestion: having thought about this a bit more, I think a different name for the secret key might be helpful.

`tls.der` has no hint that it's just the private key. From looking at the name, I might reasonably conclude that it contains a DER-encoded certificate chain. `tls.der` is arguably a better name than the already-existing k8s `tls.key` (which just provides no useful information about the format at all) but I think we can do better here.

Especially with private keys, I think it's important to be clear when something should be private. Including "key" or "privkey" or "private" somewhere in the name would help to reinforce that.

How about maybe `key.der` instead? I'd be open to other suggestions too.

seuf replied:
Good point !  `key.der` is a better name than `lts.der` . I'll make the changes.

SgtCoDFish reviewed (changes requested):
Couple more comments from me now I've had another chance to look at this!

Really appreciate your engagement on this @seuf - I love it when enthusiastic people contribute back :grin:

maelvls commented:
Hi! This PR looks very promising.

Looking at the PR description and the issue associated, I was not able to identify the concrete use-case associated with this feature. Would it be possible to add in the PR description a use-case for it? That is going to be useful when writing the release notes so that people know why this feature came in.

I suggest this:

> ### Use-case: Postgres JBDC driver (lower than 42.2.8)
>
> In recent versions of the Postgres JBDC driver (42.2.9 and above, which was released in Dec 2019), you can use the PKCS#12 format to load the private key and certificate. cert-manager supports PKCS#12 since 0.14, which means the integration with cert-manager works well.
>
> But if you are stuck with a version of the Postgres JDBC driver that is 42.2.9 and below, then the only way is to use the [`sslkey`](https://jdbc.postgresql.org/documentation/head/ssl-client.html) configuration field. The `sslkey` requires the path to a file containing the private key PKCS#8-formated and DER-encoded.

I am not a user of Postgres JBDC driver myself, please let me know if that makes sense.

seuf commented:
> I am not a user of Postgres JBDC driver myself, please let me know if that makes sense.

You're right that's exactly why I am doing this PR (we are stuck on PG 11). The version of the [libpq](https://www.postgresql.org/docs/13/libpq-connect.html#LIBPQ-CONNECT-SSLPASSWORD) must be 13+ to accept the `sslpassword` option and unlock the pkcs12 keystore.

seuf commented:
Hi, I've updated the PR description with the use case for postgresql (and a configuration example of postgresql hba).
I've also renamed `tls.der` in `key.der` :smiley:

munnerz commented:
Regarding the feature flagging - we're going to need to wire the `--feature-gates` flag into the webhook component and re-jig how we register features against different components if we want to have this all feature gated. I'll put a PR up shortly to do this, and you'll then be able to build on that more easily 😄

JoshVanL commented on pkg/controller/certificates/internal/secretsmanager/secret_test.go:
```diff
@@ -71,6 +74,15 @@ func TestSecretsManager(t *testing.T) {
+			{Type: "DER"},
+			{Type: "PEM"},
+		}),
+		gen.EnableFeatureGate(feature.ExperimentalCertificatePrivateKeyAdditionalFormats),
```
This will leak across tests. Please unset this feature gate in a testing `t.Cleanup` function.

seuf replied:
Is it ok if I disable the feature in the CheckFn of the test builder ?

munnerz replied:
Please remove the `gen.EnableFeatureGate` function (as it isn't needed from what I can tell?) and instead just use the feature util package as documented: https://github.com/jetstack/cert-manager/blob/4afe2f00a8142d0bd2134da408a811cb70e100a6/pkg/util/feature/feature_gate.go#L27

Should be:

```
defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ExperimentalCertificatePrivateKeyAdditionalFormats, true)()
```

munnerz replied:
The `defer` will take care of unsetting this after the test :)

seuf replied:
done :heavy_check_mark:

JoshVanL commented on test/e2e/framework/helper/validation/certificates/certificates.go:
```diff
@@ -37,12 +37,20 @@ import (
+		found := false
+		for _, k2 := range validKeys {
+			if k == k2 {
+				found = true
```
```suggestion
				found = true
				break
```

JoshVanL commented on test/e2e/framework/helper/validation/certificates/certificates.go:
```diff
@@ -37,12 +37,20 @@ import (
-	if !(len(secret.Data) == 2 || len(secret.Data) == 3) {
-		return fmt.Errorf("Expected 2 or 3 keys in certificate secret, but there was %d", len(secret.Data))
+// ExpectValidKeysInSecret checks that the secret contains valid keys
+func ExpectValidKeysInSecret(_ *cmapi.Certificate, secret *corev1.Secret) error {
+	validKeys := [5]string{corev1.TLSPrivateKeyKey, corev1.TLSCertKey, cmmeta.TLSCAKey, "key.der", "tls.pem"}
+	for k := range secret.Data {
+		found := false
+		for _, k2 := range validKeys {
+			if k == k2 {
+				found = true
+			}
+		}
+		if !found {
+			return fmt.Errorf("Expected secret key %s to be a valid key (%v)", k, validKeys)
+		}
 	}
```
This looks like it has changed the intention and behavior of this function. This will now return `nil`, for example, if there are _no_ keys in the Secret Data.

We should _at least_ still be catching cases where there are less or more keys than expected, not just that the keys which do exist are a member of the valid set.

seuf replied:
I agree, The test should fail if there is less thant 2 keys. Because of additional key format feature, it may have a lot of keys in secret. For example if I enable der and pem format I have 5 keys in my secret (tls.key, tls.crt, ca.crt, tls.der, tls.pem)

JoshVanL commented on test/unit/gen/certificate.go:
```diff
@@ -261,3 +264,15 @@ func SetCertificateRevisionHistoryLimit(limit int32) CertificateModifier {
+	}
+}
+
+func EnableFeatureGate(featureGate featuregate.Feature) CertificateModifier {
+	return func(crt *v1.Certificate) {
+		utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%s=%t", featureGate, true))
+	}
+}
```
This doesn't have anything to do with building a `Certificate` object, so this function doesn't belong here.

Given it only looks like it is used once, we can probably do without this helper entirely.

seuf replied:
I've added Enable/Disable feature gate function in util feature gate

JoshVanL reviewed (changes requested):
Added some review comments.

The API in its current form is a bit troubling for me, and can see it as being ambiguous or confusing. I am also not too keen on trying to overload the `keystores` field either. It will need a bit more discussion.

SgtCoDFish commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -203,8 +206,24 @@ func (s *SecretsManager) setValues(crt *cmapi.Certificate, secret *corev1.Secret
+					switch f.Type {
+					case cmapi.DERKeyFormat:
+						block, _ := pem.Decode(data.PrivateKey)
+						secret.Data["key.der"] = block.Bytes
+					case cmapi.PEMKeyFormat:
+						secret.Data["tls.pem"] = []byte(fmt.Sprintf("%s%s", data.PrivateKey, data.Certificate))
```
The equivalent keys for keystore types aren't exported in the API:

https://github.com/jetstack/cert-manager/blob/7923823cff0afe5af49282bde3a5a1cc6f5abc84/pkg/controller/certificates/internal/secretsmanager/keystore.go#L37-L49

Making the new `"key.der"` constant and exported in the API would be a nice-to-have but I don't think is required in this PR given the keystore equivalents aren't exported. They'd go under `cert-manager.io`.

seuf commented on internal/apis/certmanager/types_certificate.go:
```diff
@@ -203,6 +203,25 @@ type CertificatePrivateKey struct {
+	AdditionalKeyFormats []AdditionalKeyFormat
+}
+
+type KeyFormat string
```
Yes, but as there is a PEM format which can containt both key and cert I thought keyformat was ok.
I make the change

seuf commented on test/unit/gen/certificate.go:
```diff
@@ -261,3 +264,15 @@ func SetCertificateRevisionHistoryLimit(limit int32) CertificateModifier {
+
+func SetCertificateAdditionalKeyFormats(additionalKeyFormats []v1.AdditionalKeyFormat) CertificateModifier {
+	return func(crt *v1.Certificate) {
+		crt.Spec.PrivateKey.AdditionalKeyFormats = additionalKeyFormats
```
mmh.. not sure.. the `SetCertificateKeySize` and `SetCertificateKeyAlgorithm` functions are not doing this..

munnerz replied:
That's a good catch too @seuf - we should be doing this everywhere to avoid a nil pointer dereference :) (although we should hopefully only ever see these during test cases, so errors will be picked up long before a user gets this code!)

seuf replied:
The additionalOutputFomat is now at the root of the spec. So no need to check if privateKey is not nil here. (It should be test in other functions, but that's another issue :yum: )

seuf commented on pkg/feature/features.go:
```diff
@@ -40,6 +40,11 @@ const (
+	// alpha: v1.7.0
+	//
+	// ExperimentalCertificatePrivateKeyAdditionalFormats enable private key additional format
+	ExperimentalCertificatePrivateKeyAdditionalFormats featuregate.Feature = "ExperimentalCertificatePrivateKeyAdditionalFormats"
```
what dou you mean by graduation path ?

munnerz replied:
Despite the others _not_ being named accordingly, I think we're better to not have `Experimental` in the name here. Maybe just `AdditionalCertificateOutputFormats`? (given that we're expecting this feature to also cover more than just private keys too).

> what dou you mean by graduation path ?

I think Josh refers to how we make this feature go from alpha->beta->GA (and as part of that, enable the feature by default). I think this is a good one to discuss at the biweekly community meeting on Wednesday's. It may just be that we note someone here as an 'owner' of the feature, and that person is responsible for 'pushing the feature forward' once this PR has merged and the feature has been rolled out to users :)

seuf replied:
OK, So I've renamed the feature to `AdditionalCertificateOutputFormats`. I'll try to connect at the next community meeting to discuss about the graduation path.

seuf commented:
Hi @JoshVanL I've implemented your change requests.

seuf commented on internal/apis/certmanager/validation/certificate.go:
```diff
@@ -77,6 +79,11 @@ func ValidateCertificateSpec(crt *internalcmapi.CertificateSpec, fldPath *field.
 		default:
 			el = append(el, field.Invalid(fldPath.Child("privateKey", "algorithm"), crt.PrivateKey.Algorithm, "must be either empty or one of rsa or ecdsa"))
 		}
+		if crt.PrivateKey.AdditionalKeyFormats != nil {
```
+1

munnerz commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -231,6 +231,10 @@ type CertificatePrivateKey struct {
+
+	// Other formats in which the private key should be written to the secret
+	// +optional
+	AdditionalKeyFormats []AdditionalKeyFormat `json:"additionalKeyFormats,omitempty"`
```
I think this should be named `additionalOutputFormats` and set at the root of the `Spec` struct (e.g. `spec.additionalOutputFormats`).

Otherwise I don't think it makes sense for us to support e.g. `CombinedPEM` here in future, as the positioning of it _today_ implies that it's only an additional format for the *privateKey*, which isn't true with CombinedPEM

seuf replied:
As discussed on slack we agreed to use the `additionalOutputFormats` in the `CertificateSpec` .

munnerz commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,29 @@ var (
+// When Type is set to `CombinedPEM` and additional entry `tls.pem` will be created in the secret
+// containing the PEM formated certificate (tls.key + tls.crt concatened)
+// +kubebuilder:validation:Enum=DER;CombinedPEM
+type PrivateKeyFormat string
```
`type OutputFormat string`

seuf replied:
:heavy_check_mark:

munnerz commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,29 @@ var (
+	PrivateKeyFormatPEM PrivateKeyFormat = "CombinedPEM"
+)
+
+type AdditionalKeyFormat struct {
```
`type OutputFormat struct`

seuf replied:
renamed field

munnerz commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -203,8 +206,24 @@ func (s *SecretsManager) setValues(crt *cmapi.Certificate, secret *corev1.Secret
+						block, _ := pem.Decode(data.PrivateKey)
+						secret.Data["key.der"] = block.Bytes
+					case cmapi.PrivateKeyFormatPEM:
+						secret.Data["tls.pem"] = []byte(fmt.Sprintf("%s%s", data.PrivateKey, data.Certificate))
```
Are we sure we'll handle new-lines here correctly? `strings.Join([]string{data.PrivateKey, data.Certificate}, "\n")` may ensure we handle this more correctly in edge-cases where the private key doesn't end with a newline :)

seuf replied:
Using String.join instead of fmt.Sprintf with "\n" at the end to be sure.

munnerz commented on pkg/controller/certificates/internal/secretsmanager/secret_test.go:
```diff
@@ -71,6 +75,16 @@ func TestSecretsManager(t *testing.T) {
 		}),
 	)
+	// enable feature gate additional private key for this test
+	utilfeature.EnableFeatureGate(feature.ExperimentalCertificatePrivateKeyAdditionalFormats)
```
Please use `defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()` so we can ensure this is *unset* again after the test

seuf replied:
done :heavy_check_mark:

munnerz commented on pkg/util/feature/feature_gate.go:
```diff
@@ -31,3 +33,13 @@ var (
+	return nil
+}
+
+func DisableFeatureGate(featureGate featuregate.Feature) error {
```
Please remove both of these functions - they aren't needed

seuf replied:
ok, now using SetFeatureGateDuringTest instead

seuf commented on test/e2e/suite/issuers/ca/certificate.go:
```diff
@@ -124,6 +124,27 @@ var _ = framework.CertManagerDescribe("CA Certificate", func() {
+
+			cert := util.NewCertManagerBasicCertificate(certificateName, certificateSecretName, issuerName, v1.IssuerKind, nil, nil)
+			cert.Spec.PrivateKey.AdditionalKeyFormats = []v1.AdditionalKeyFormat{
+				{Type: "DER"},
```
added `CombinedPem` test.

SgtCoDFish commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,37 @@ var (
+	// stored in `tls.pem` secret key
+	OutputFormatPEM OutputFormatType = "CombinedPEM"
+
+	// OutputFormatPEMKey is the name of the data entry in the Secret resource
+	// used to store the Combined PEM certifcate + key
+	OutputFormatPEMKey string = "tls.pem"
```
suggestion: I feel like `tls.pem` really needs a different name. Both `tls.crt` and `tls.key` in a k8s TLS secret contain a PEM-encoded value, so `tls.pem` could work as a name for both of them. It runs the risk of being confusing.

I'd prefer something verbose but more clear such as `tls-combined.pem` or `tls-chain-key-combined.pem`. I'm not _thrilled_ with either of those suggestions, but I really worry that `tls.pem` isn't clear enough.

seuf replied:
OK, i change to `tls-combined.pem`

SgtCoDFish commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,37 @@ var (
+	OutputFormatPEM OutputFormatType = "CombinedPEM"
+
+	// OutputFormatPEMKey is the name of the data entry in the Secret resource
+	// used to store the Combined PEM certifcate + key
```
typo:
```suggestion
	// used to store the Combined PEM certificate + key
```

seuf replied:
oops, fixed

SgtCoDFish commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,37 @@ var (
 )
 
+// The type of additional private key format to write
+// Allowed values are `DER` or `PEM`.
```
suggestion:

```suggestion
// Allowed values are `DER` or `CombinedPEM`.
```

seuf replied:
right, fixed

SgtCoDFish commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,37 @@ var (
+// Allowed values are `DER` or `PEM`.
+// When Type is set to `DER` an additional entry `key.der` will be created in the secret
+// containing the binary format of the key.
+// When Type is set to `CombinedPEM` and additional entry `tls.pem` will be created in the secret
+// containing the PEM formated certificate (tls.key + tls.crt concatened)
```
three typos:

```suggestion
// When Type is set to `CombinedPEM` an additional entry `tls.pem` will be created in the secret
// containing the PEM formatted certificate (tls.key + tls.crt concatenated)
```

seuf replied:
thanks for spotting all the mistakes :sweat_smile:

SgtCoDFish commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -120,6 +123,40 @@ func (s *SecretsManager) UpdateData(ctx context.Context, crt *cmapi.Certificate,
 	return err
 }
 
+func updateSecretWithAdditionalOutputFormats(crt *cmapi.Certificate, secret *corev1.Secret, data SecretData) error {
+	if crt.Spec.AdditionalOutputFormats != nil {
+		additionalOutputFormatDER := false
+		additionalOutputFormatPEM := false
+		for _, f := range crt.Spec.AdditionalOutputFormats {
+			switch f.Type {
+			case cmapi.OutputFormatDER:
+				additionalOutputFormatDER = true
+			case cmapi.OutputFormatPEM:
+				additionalOutputFormatPEM = true
+			default:
+				return fmt.Errorf("Unknown additional output format %s", f.Type)
+			}
+		}
+		if additionalOutputFormatDER {
+			// Store binary format of the private key
+			block, _ := pem.Decode(data.PrivateKey)
+			secret.Data[cmapi.OutputFormatDERKey] = block.Bytes
+		} else {
+			delete(secret.Data, cmapi.OutputFormatDERKey)
+		}
+		if additionalOutputFormatPEM {
+			// Combine tls.key and tls.crt
+			secret.Data[cmapi.OutputFormatPEMKey] = []byte(strings.Join([]string{string(data.PrivateKey), string(data.Certificate)}, "\n"))
+		} else {
+			delete(secret.Data, cmapi.OutputFormatPEMKey)
+		}
+	} else {
+		delete(secret.Data, cmapi.OutputFormatDERKey)
+		delete(secret.Data, cmapi.OutputFormatPEMKey)
+	}
+	return nil
+}
```
suggestion: I'd usually prefer to exit early if `crt.Spec.AdditionalOutputFormats == nil` and I think the code is more readable and idiomatic that way.

```suggestion
func updateSecretWithAdditionalOutputFormats(crt *cmapi.Certificate, secret *corev1.Secret, data SecretData) error {
	if crt.Spec.AdditionalOutputFormats == nil {
		delete(secret.Data, cmapi.OutputFormatDERKey)
		delete(secret.Data, cmapi.OutputFormatPEMKey)
		return nil
	}

	additionalOutputFormatDER := false
	additionalOutputFormatPEM := false

	for _, f := range crt.Spec.AdditionalOutputFormats {
		switch f.Type {
		case cmapi.OutputFormatDER:
			additionalOutputFormatDER = true
		case cmapi.OutputFormatPEM:
			additionalOutputFormatPEM = true
		default:
			return fmt.Errorf("Unknown additional output format %s", f.Type)
		}
	}

	if additionalOutputFormatDER {
		// Store binary format of the private key
		block, _ := pem.Decode(data.PrivateKey)
		secret.Data[cmapi.OutputFormatDERKey] = block.Bytes
	} else {
		delete(secret.Data, cmapi.OutputFormatDERKey)
	}

	if additionalOutputFormatPEM {
		// Combine tls.key and tls.crt
		secret.Data[cmapi.OutputFormatPEMKey] = []byte(strings.Join([]string{string(data.PrivateKey), string(data.Certificate)}, "\n"))
	} else {
		delete(secret.Data, cmapi.OutputFormatPEMKey)
	}

	return nil
}
```

seuf replied:
ok, I've exited early like suggested

SgtCoDFish commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,37 @@ var (
 	RotationPolicyAlways PrivateKeyRotationPolicy = "Always"
 )
 
+// The type of additional private key format to write
```
suggestion: This line doesn't read right when it's serialized into the description for `deploy/crds/crd-certificates.yaml`. 
```suggestion
// The type of additional private key format to write.
```

seuf replied:
added dot `.`

SgtCoDFish reviewed (dismissed):
Thank you for your persistence with this PR, it's much appreciated :grin: 

I've got a few comments; I've not had a really deep look at this but there are a few typos and things which I think might be worth addressing.

(Comments on `pkg/apis/certmanager/v1/types_certificate.go` apply to the other versions too)

seuf commented:
Hi @SgtCoDFish I've updated the PR with fix you suggested. Thanks again for the review.
I've just tested it in my cluster and it works as expected (tested both CombinedPem and DER formats) :smiley:

SgtCoDFish commented:
> Hi @SgtCoDFish I've updated the PR with fix you suggested. Thanks again for the review. I've just tested it in my cluster and it works as expected (tested both CombinedPem and DER formats) smiley

Thanks @seuf - I've taken another look and I've got a commit ready to fix a couple of other typos and suggest some name changes.

If you give me permission to push to your fork of cert-manager, I can push directly. Otherwise, I've attached a patch you could use with `git apply`.

[spelling-rename-etc.txt](https://github.com/jetstack/cert-manager/files/7848824/spelling-rename-etc.txt)

seuf commented:
@SgtCoDFish I've git applied your changed (and squashed the commit). 
I've also send an invite to my forked repo if you want to add more changes.

JoshVanL commented on internal/apis/certmanager/v1alpha2/types_certificate.go:
```diff
@@ -208,6 +208,10 @@ type CertificateSpec struct {
 	// +kubebuilder:validation:ExclusiveMaximum=false
 	// +optional
```
Guess this branch needs to be rebased since `v1alpha2` doesn't exist anymore?

seuf replied:
I think it shouldn't exists anymore but it's still present on the master branch.
If we run the `./hack/update-codegen.sh` script from the master branch it remove a lot of conversion / deepcopy functions (and it doe not build anymore).
That's not the goal of this branch.

JoshVanL replied:
Thanks @seuf and sorry for the confusion here. I had not appreciated the fact that the old APIs had moved under `internal`, rather than being deleted completely.

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -187,6 +187,10 @@ type CertificateSpec struct {
 	// +kubebuilder:validation:ExclusiveMaximum=false
 	// +optional
 	RevisionHistoryLimit *int32 `json:"revisionHistoryLimit,omitempty"` // Validated by the validating webhook.
+
```
nit: wrap lines and add full stop at the ned (since this is exported documentation on the output API/CRD.

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
 	RotationPolicyAlways PrivateKeyRotationPolicy = "Always"
 )
 
+// OutputFormatType specifies which additional key formats should be added to Kubernetes secrets.
```
```suggestion
// OutputFormatType specifies which additional output formats should be added to Kubernetes secrets.
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+type OutputFormatType string
+
+const (
+	// AdditionalKeyOutputFormatDER requests that the DER binary format of the key
```
```suggestion
	// AdditionalOutputFormatDER requests that the DER binary format of the key
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+
+const (
+	// AdditionalKeyOutputFormatDER requests that the DER binary format of the key
+	// is stored in the `key.der` key of a secret
```
```suggestion
	// is stored in the `key.der` key of the Certificate's Secret.
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+const (
+	// AdditionalKeyOutputFormatDER requests that the DER binary format of the key
+	// is stored in the `key.der` key of a secret
+	AdditionalKeyOutputFormatDER OutputFormatType = "DER"
```
The prefix is inconsistent here with the other `OutputFormatType` and the type name itself.

```suggestion
	AdditionalOutputFormatDER OutputFormatType = "DER"
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+	AdditionalOutputFormatPEMKey string = "tls-combined.pem"
+)
+
+// AdditionalOutputFormat wraps an additional key output format type
```
```suggestion
// AdditionalOutputFormat wraps an additional key output format type.
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+
+// AdditionalOutputFormat wraps an additional key output format type
+type AdditionalOutputFormat struct {
+	Type OutputFormatType `json:"type"`
```
Add comment to the field here since it will be exported into docs for website/CRD.

munnerz replied:
Let's follow up with another PR to fix this 😊

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+
+	// AdditionalKeyOutputFormatDERKey is the name of the data entry in the Secret resource
+	// used to store the DER formatted private key
+	AdditionalKeyOutputFormatDERKey string = "key.der"
```
Same with prefix inconsistency. 

```suggestion
	AdditionalOutputFormatDERKey string = "key.der"
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+	// is stored in the `key.der` key of a secret
+	AdditionalKeyOutputFormatDER OutputFormatType = "DER"
+
+	// AdditionalKeyOutputFormatDERKey is the name of the data entry in the Secret resource
+	// used to store the DER formatted private key
```
```suggestion
	// AdditionalKeyOutputFormatDERKey is the data key entry in the Secret resource
	// used to store the DER formatted private key.
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+	// used to store the DER formatted private key
+	AdditionalKeyOutputFormatDERKey string = "key.der"
+
+	// AdditionalOutputFormatCombinedPEM requests that an entry containing both the PEM tls.crt and tls.key
+	// values concatenated is stored in the `tls-combined.pem` key of a secret
```
```suggestion
	// AdditionalOutputFormatCombinedPEM requests that an entry containing both the PEM tls.crt and tls.key
	// values concatenated is stored in the `tls-combined.pem` key of the Certificate's Secret.
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+	// values concatenated is stored in the `tls-combined.pem` key of a secret
+	AdditionalOutputFormatCombinedPEM OutputFormatType = "CombinedPEM"
+
+	// AdditionalOutputFormatPEMKey is the name of the data entry in the Secret resource
+	// used to store the combined PEM certificate + key
```
```suggestion
	// AdditionalKeyOutputFormatDERKey is the data key entry in the Secret resource
	// used to store the combined PEM certificate concatenated with the private key.
```

JoshVanL commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -120,6 +123,45 @@ func (s *SecretsManager) UpdateData(ctx context.Context, crt *cmapi.Certificate,
 }
 
+func updateSecretWithAdditionalOutputFormats(crt *cmapi.Certificate, secret *corev1.Secret, data SecretData) error {
+	if crt.Spec.AdditionalOutputFormats == nil {
+		delete(secret.Data, cmapi.AdditionalKeyOutputFormatDERKey)
+		delete(secret.Data, cmapi.AdditionalOutputFormatPEMKey)
+		return nil
+	}
+
+	additionalOutputFormatDER := false
+	additionalOutputFormatPEM := false
+
+	for _, f := range crt.Spec.AdditionalOutputFormats {
+		switch f.Type {
+		case cmapi.AdditionalKeyOutputFormatDER:
+			additionalOutputFormatDER = true
+		case cmapi.AdditionalOutputFormatCombinedPEM:
+			additionalOutputFormatPEM = true
+		default:
+			return fmt.Errorf("Unknown additional output format %s", f.Type)
+		}
+	}
+
+	if additionalOutputFormatDER {
+		// Store binary format of the private key
+		block, _ := pem.Decode(data.PrivateKey)
+		secret.Data[cmapi.AdditionalKeyOutputFormatDERKey] = block.Bytes
+	} else {
+		delete(secret.Data, cmapi.AdditionalKeyOutputFormatDERKey)
+	}
+
+	if additionalOutputFormatPEM {
+		// Combine tls.key and tls.crt
+		secret.Data[cmapi.AdditionalOutputFormatPEMKey] = []byte(strings.Join([]string{string(data.PrivateKey), string(data.Certificate)}, "\n"))
+	} else {
+		delete(secret.Data, cmapi.AdditionalOutputFormatPEMKey)
+	}
+
+	return nil
+}
```
This is fine as is, but heads up that this will look a bit different after https://github.com/jetstack/cert-manager/pull/4638

JoshVanL commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -120,6 +123,45 @@ func (s *SecretsManager) UpdateData(ctx context.Context, crt *cmapi.Certificate,
+		case cmapi.AdditionalOutputFormatCombinedPEM:
+			additionalOutputFormatPEM = true
+		default:
+			return fmt.Errorf("Unknown additional output format %s", f.Type)
```
```suggestion
			return fmt.Errorf("unknown additional output format %s", f.Type)
```

JoshVanL commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -120,6 +123,45 @@ func (s *SecretsManager) UpdateData(ctx context.Context, crt *cmapi.Certificate,
+
+	if additionalOutputFormatPEM {
+		// Combine tls.key and tls.crt
+		secret.Data[cmapi.AdditionalOutputFormatPEMKey] = []byte(strings.Join([]string{string(data.PrivateKey), string(data.Certificate)}, "\n"))
```
```suggestion
		secret.Data[cmapi.AdditionalOutputFormatPEMKey] = bytes.Join([]byte{data.PrivateKey, data.Certificate}, []byte("\n"))
```

munnerz replied:
Can you follow up with this?

JoshVanL commented on pkg/controller/certificates/internal/secretsmanager/secret.go:
```diff
@@ -203,8 +245,12 @@ func (s *SecretsManager) setValues(crt *cmapi.Certificate, secret *corev1.Secret
 
+		// Add additional output formats
+		if utilfeature.DefaultFeatureGate.Enabled(feature.AdditionalCertificateOutputFormats) {
+			updateSecretWithAdditionalOutputFormats(crt, secret, data)
```
Check the error.

JoshVanL replied:
We need to be checking the error returned from `updateSecretWithAdditionalOutputFormats` here.

JoshVanL commented on pkg/controller/certificates/internal/secretsmanager/secret_test.go:
```diff
@@ -400,6 +414,47 @@ func TestSecretsManager(t *testing.T) {
 			},
 			expectedErr: false,
 		},
+
```
Please add a test for:

"if secret exists with 'tls-combined.pem' and 'tls.der' set, but Certificate has no output formats set, update Secret removing those fields".

We should also have the combination of the above with 2 more tests for removing _just_ `tls-combined.pem'  or just `tls.der`.

seuf replied:
OK, I've added more tests. Now we have
- if secret does not exists and we want additional output format DER only
- if secret does not exists and we want additional output format CombinedPEM only
- if secret does not exists and we want additional output format DER and CombinedPEM
- if secret exists with tls-combined.pem and der.key and we don't want any additional output format
- if secret exists with tls-combined.pem and der.key and we only want DER additional output format
- if secret exists with tls-combined.pem and der.key and we only want CombinedPEM additional output format

JoshVanL commented on test/e2e/framework/helper/validation/certificates/certificates.go:
```diff
@@ -37,12 +37,26 @@ import (
+			}
+		}
+		if !found {
+			return fmt.Errorf("Expected secret key %s to be a valid key (%v)", k, validKeys)
```
```suggestion
			return fmt.Errorf("expected secret key %q to be a valid key (%v)", k, validKeys)
```

JoshVanL replied:
I'm not sure this find check here is right.. We don't enforce on Certificate secrets that _only_ valid keys can be present, and fully support users adding there own keys. We should remove this found check.

seuf replied:
I Agree, users can add additional keys in kubernetes secret via kustomize secretGenerator or whatever.
I've removed the "found" test.

JoshVanL commented on test/e2e/suite/issuers/ca/certificate.go:
```diff
@@ -124,6 +124,27 @@ var _ = framework.CertManagerDescribe("CA Certificate", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Validating the issued Certificate...")
+			err = f.Helper().ValidateCertificate(cert)
```
This comment is still valid. I can't see a validation check where _if_ a output format is specified, the contents on the resulting value is checked against what we expect (`combined-pem.tls` really does contain the concatenated certificate and key, and `der.tls` really does contain the private key in DER format).

JoshVanL commented on test/unit/gen/certificate.go:
```diff
@@ -261,3 +261,9 @@ func SetCertificateRevisionHistoryLimit(limit int32) CertificateModifier {
 	}
 }
+
+func SetCertificateAdditionalOutputFormats(additionalOutputFormats []v1.AdditionalOutputFormat) CertificateModifier {
+	return func(crt *v1.Certificate) {
+		crt.Spec.AdditionalOutputFormats = additionalOutputFormats
+	}
+}
```
This is a good candidate for a variadic function like used elsewhere in these generators.

```suggestion
func SetCertificateAdditionalOutputFormats(additionalOutputFormats ...v1.AdditionalOutputFormat) CertificateModifier {
	return func(crt *v1.Certificate) {
		crt.Spec.AdditionalOutputFormats = additionalOutputFormats
	}
}
```

JoshVanL commented on pkg/apis/certmanager/v1/types_certificate.go:
```diff
@@ -249,6 +253,38 @@ var (
+type AdditionalOutputFormat struct {
+	Type OutputFormatType `json:"type"`
+}
+
```
We should copy these back to internal if/when the above are changed.

JoshVanL replied:
We should copy these back to internal if/when the above are changed.

JoshVanL commented on deploy/crds/crd-certificates.yaml:
```diff
@@ -67,6 +67,21 @@ spec:
                 - secretName
               properties:
+                additionalOutputFormats:
+                  description: AdditionalOutputFormats allows for requests for additional formats in which the private key should be written to the secret
```
We should make a note in the comments of the API (so it is displayed here) that this feature is in Alpha, and if the user wants to use it, they will need to enable the feature gate else it won't do anything.

seuf replied:
Agree, comment added.

JoshVanL reviewed (changes requested):
Added some more comments :)

Since it looks like we are going ahead with this feature, we should open up a website PR documenting this feature (and gate).

seuf commented:
I've updated comments variables and added more unit tests as you suggested.
Regarding the e2e tests, the content of the `tls-combined.pem` or `key.der` are not verified. This is done in unit tests.

JoshVanL commented:
Appreciate your persistence on this @seuf, I know it has been a bit painful getting this over the line :smile: 

Thank you :pray: 

/lgtm
/approve

aarononeal commented:
It would be nice if there were an annotation or a global config default so this could work for `Ingress` resources as well.
